# Exercício Gráfico Fluxo.
Vocês vão utilizar um dos tipos de gráfico do tipo Fluxo (Flow).  
  
Referência: https://www.python-graph-gallery.com/  
  
Em grupo vocês devem decidir qual base aberta vão utilizar para fazer uma demonstracão de um fluxo de dados.  
  
No final da aula de hoje vocês devem ter decidido qual base utilizar e se possível uma tentatativa de como utilizar um dos gráficos de Fluxo.  
  

In [69]:
# !pip install plotly
# !pip install colorlover
import pandas as pd
import numpy as np
import plotly.graph_objs as go
import plotly
import chart_studio.plotly as py
from plotly.offline import download_plotlyjs, init_notebook_mode, iplot, plot
#import colorlover as cl

In [3]:
df = pd.read_csv('./trip.csv', sep=',')
df = df.drop_duplicates(subset=['trip_id'])

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 236065 entries, 0 to 286857
Data columns (total 12 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   trip_id            236065 non-null  int64  
 1   starttime          236065 non-null  object 
 2   stoptime           236065 non-null  object 
 3   bikeid             236065 non-null  object 
 4   tripduration       236065 non-null  float64
 5   from_station_name  236065 non-null  object 
 6   to_station_name    236065 non-null  object 
 7   from_station_id    236065 non-null  object 
 8   to_station_id      236065 non-null  object 
 9   usertype           236064 non-null  object 
 10  gender             146171 non-null  object 
 11  birthyear          146167 non-null  float64
dtypes: float64(2), int64(1), object(9)
memory usage: 23.4+ MB


In [5]:
df['from_station_name'].unique()

array(['2nd Ave & Spring St',
       'Occidental Park / Occidental Ave S & S Washington St',
       'City Hall / 4th Ave & James St',
       'King Street Station Plaza / 2nd Ave Extension S & S Jackson St',
       '1st Ave & Marion St', 'E Harrison St & Broadway Ave E',
       '9th Ave N & Mercer St',
       'Burke-Gilman Trail / NE Blakeley St & 24th Ave NE',
       'Seattle Aquarium / Alaskan Way S & Elliott Bay Trail',
       '6th Ave & Blanchard St',
       'Burke Museum / E Stevens Way NE & Memorial Way NE',
       '2nd Ave & Vine St', '6th Ave S & S King St',
       'NE 42nd St & University Way NE', '12th Ave & NE Campus Pkwy',
       '2nd Ave & Pine St', 'Frye Art Museum / Terry Ave & Columbia St',
       'Harvard Ave & E Pine St', 'Westlake Ave & 6th Ave',
       'Dexter Ave & Denny Way',
       'UW Magnuson Health Sciences Center Rotunda / Columbia Rd & San Juan Rd',
       '12th Ave & E Mercer St', '2nd Ave & Blanchard St',
       'REI / Yale Ave N & John St', 'NE 47th St & 1

In [6]:
df[['from_station_name','to_station_name']]

from_station_name  \
0                          2nd Ave & Spring St   
1                          2nd Ave & Spring St   
2                          2nd Ave & Spring St   
3                          2nd Ave & Spring St   
4                          2nd Ave & Spring St   
...                                        ...   
286853                 Harvard Ave & E Pine St   
286854  Cal Anderson Park / 11th Ave & Pine St   
286855  Cal Anderson Park / 11th Ave & Pine St   
286856  Cal Anderson Park / 11th Ave & Pine St   
286857  Cal Anderson Park / 11th Ave & Pine St   

                                          to_station_name  
0       Occidental Park / Occidental Ave S & S Washing...  
1       Occidental Park / Occidental Ave S & S Washing...  
2       Occidental Park / Occidental Ave S & S Washing...  
3       Occidental Park / Occidental Ave S & S Washing...  
4       Occidental Park / Occidental Ave S & S Washing...  
...                                                   ...  
286853                                2nd Ave & Spring St  
286854                              6th Ave S & S King St  
286855                              6th Ave S & S King St  
286856                              6th Ave S & S King St  
286857                              6th Ave S & S King St  

[236065 rows x 2 columns]

In [7]:
crosstab_df = pd.crosstab(df['from_station_name'], df['to_station_name'])

In [8]:
crosstab_df

to_station_name                                     12th Ave & E Denny Way  \
from_station_name                                                            
12th Ave & E Denny Way                                                  66   
12th Ave & E Mercer St                                                 134   
12th Ave & E Yesler Way                                                 21   
12th Ave & NE Campus Pkwy                                                5   
15th Ave E & E Thomas St                                                40   
...                                                                    ...   
UW Intramural Activities Building                                        1   
UW Magnuson Health Sciences Center Rotunda / Co...                       0   
UW McCarty Hall / Whitman Ct                                             1   
Union St & 4th Ave                                                       2   
Westlake Ave & 6th Ave                                                   8   

to_station_name                                     12th Ave & E Mercer St  \
from_station_name                                                            
12th Ave & E Denny Way                                                 117   
12th Ave & E Mercer St                                                 199   
12th Ave & E Yesler Way                                                 24   
12th Ave & NE Campus Pkwy                                               14   
15th Ave E & E Thomas St                                               107   
...                                                                    ...   
UW Intramural Activities Building                                        8   
UW Magnuson Health Sciences Center Rotunda / Co...                       3   
UW McCarty Hall / Whitman Ct                                             1   
Union St & 4th Ave                                                       7   
Westlake Ave & 6th Ave                                                  12   

to_station_name                                     12th Ave & E Yesler Way  \
from_station_name                                                             
12th Ave & E Denny Way                                                   55   
12th Ave & E Mercer St                                                   32   
12th Ave & E Yesler Way                                                  44   
12th Ave & NE Campus Pkwy                                                 4   
15th Ave E & E Thomas St                                                 27   
...                                                                     ...   
UW Intramural Activities Building                                         2   
UW Magnuson Health Sciences Center Rotunda / Co...                        0   
UW McCarty Hall / Whitman Ct                                              0   
Union St & 4th Ave                                                       16   
Westlake Ave & 6th Ave                                                    4   

to_station_name                                     12th Ave & NE Campus Pkwy  \
from_station_name                                                               
12th Ave & E Denny Way                                                     15   
12th Ave & E Mercer St                                                     64   
12th Ave & E Yesler Way                                                     6   
12th Ave & NE Campus Pkwy                                                 655   
15th Ave E & E Thomas St                                                  183   
...                                                                       ...   
UW Intramural Activities Building                                         150   
UW Magnuson Health Sciences Center Rotunda / Co...                         78   
UW McCarty Hall / Whitman Ct                                               26   
Union St & 4th Ave                                 

In [9]:
top_10_estacoes = df.groupby(['from_station_name', 'to_station_name']).size().to_frame().nlargest(10, 0).reset_index()

In [10]:
top_10_estacoes

from_station_name  \
0                    Pier 69 / Alaskan Way & Clay St   
1                    Pier 69 / Alaskan Way & Clay St   
2  Seattle Aquarium / Alaskan Way S & Elliott Bay...   
3  Seattle Aquarium / Alaskan Way S & Elliott Bay...   
4                                 3rd Ave & Broad St   
5                                  2nd Ave & Pine St   
6                    Pier 69 / Alaskan Way & Clay St   
7          Lake Union Park / Valley St & Boren Ave N   
8                               E Pine St & 16th Ave   
9  Burke-Gilman Trail / NE Blakeley St & 24th Ave NE   

                                     to_station_name     0  
0                    Pier 69 / Alaskan Way & Clay St  4209  
1  Seattle Aquarium / Alaskan Way S & Elliott Bay...  2291  
2                    Pier 69 / Alaskan Way & Clay St  1605  
3  Seattle Aquarium / Alaskan Way S & Elliott Bay...  1150  
4                                  2nd Ave & Pine St  1101  
5                                 3rd Ave & Broad St   962  
6  King Street Station Plaza / 2nd Ave Extension ...   962  
7          Lake Union Park / Valley St & Boren Ave N   960  
8                                  Pine St & 9th Ave   934  
9  Burke-Gilman Trail / NE Blakeley St & 24th Ave NE   880

In [11]:
from_station_mask = top_10_estacoes['from_station_name'].unique()
to_station_mask = top_10_estacoes['to_station_name'].unique()

In [12]:
set(from_station_mask)

{'2nd Ave & Pine St',
 '3rd Ave & Broad St',
 'Burke-Gilman Trail / NE Blakeley St & 24th Ave NE',
 'E Pine St & 16th Ave',
 'Lake Union Park / Valley St & Boren Ave N',
 'Pier 69 / Alaskan Way & Clay St',
 'Seattle Aquarium / Alaskan Way S & Elliott Bay Trail'}

In [13]:
to_station_mask

array(['Pier 69 / Alaskan Way & Clay St',
       'Seattle Aquarium / Alaskan Way S & Elliott Bay Trail',
       '2nd Ave & Pine St', '3rd Ave & Broad St',
       'King Street Station Plaza / 2nd Ave Extension S & S Jackson St',
       'Lake Union Park / Valley St & Boren Ave N', 'Pine St & 9th Ave',
       'Burke-Gilman Trail / NE Blakeley St & 24th Ave NE'], dtype=object)

In [14]:
[*to_station_mask, *from_station_mask]

['Pier 69 / Alaskan Way & Clay St',
 'Seattle Aquarium / Alaskan Way S & Elliott Bay Trail',
 '2nd Ave & Pine St',
 '3rd Ave & Broad St',
 'King Street Station Plaza / 2nd Ave Extension S & S Jackson St',
 'Lake Union Park / Valley St & Boren Ave N',
 'Pine St & 9th Ave',
 'Burke-Gilman Trail / NE Blakeley St & 24th Ave NE',
 'Pier 69 / Alaskan Way & Clay St',
 'Seattle Aquarium / Alaskan Way S & Elliott Bay Trail',
 '3rd Ave & Broad St',
 '2nd Ave & Pine St',
 'Lake Union Park / Valley St & Boren Ave N',
 'E Pine St & 16th Ave',
 'Burke-Gilman Trail / NE Blakeley St & 24th Ave NE']

In [15]:
union_stations = set([*from_station_mask, *to_station_mask])

In [16]:
union_stations

{'2nd Ave & Pine St',
 '3rd Ave & Broad St',
 'Burke-Gilman Trail / NE Blakeley St & 24th Ave NE',
 'E Pine St & 16th Ave',
 'King Street Station Plaza / 2nd Ave Extension S & S Jackson St',
 'Lake Union Park / Valley St & Boren Ave N',
 'Pier 69 / Alaskan Way & Clay St',
 'Pine St & 9th Ave',
 'Seattle Aquarium / Alaskan Way S & Elliott Bay Trail'}

In [17]:
matrix_data = crosstab_df.loc[union_stations,union_stations]

In [18]:
matrix = matrix_data.copy()

name_dict = {'Burke-Gilman Trail / NE Blakeley St & 24th Ave NE': 'Burke-Gilman Trail',
                                 'King Street Station Plaza / 2nd Ave Extension S & S Jackson St': 'King Street Station Plaza',
                                 'Lake Union Park / Valley St & Boren Ave N':'Lake Union Park',
                                 'Pier 69 / Alaskan Way & Clay St':'Pier 69',
                                 'Seattle Aquarium / Alaskan Way S & Elliott Bay Trail':'Seattle Aquarium'}

matrix.rename(columns = name_dict, index = name_dict, inplace = True)

In [332]:
#def get_spaced_colors(n, randomized=False):
#    if n > 0:
#        max_value = 255
#        interval = max_value / n
#        hues = np.arange(0, max_value, interval)
#        return cl.to_rgb(["hsl(%d,80%%,40%%)" % i for i in hues])
#    else:
#        return None


PI = np.pi


def check_square(M):
    d, n = M.shape
    if d != n:
        raise ValueError("Data array must be square.")
    return n


def moduloAB(x, a, b):
    if a >= b:
        raise ValueError("Incorrect inverval ends")
    y = (x - a) % (b - a)
    return y + b if y < 0 else y + a


def test_2PI(x):
    return 0 <= x < 2 * PI


def get_ideogram_ends(ideaogram_len, gap):
    ideo_ends = []
    left = 0
    for k in range(len(ideaogram_len)):
        right = left + ideaogram_len[k]
        ideo_ends.append([left, right])
        left = right + gap
    return ideo_ends


def make_ideogram_arc(R, phi, a=50):
    # R is the circle radius
    # Phi is a list of the ends angle coordinates of an arc
    # a is a parameter that controls the number of points to be evaluated
    if not test_2PI(phi[0]) or not test_2PI(phi[1]):
        phi = [moduloAB(t, 0, 2 * PI) for t in phi]
    length = (phi[1] - phi[0]) % 2 * PI
    nr = 5 if length <= PI / 4 else int(a * length / PI)
    if phi[0] < phi[1]:
        nr = 100

        theta = np.linspace(phi[0], phi[1], nr)
    else:
        phi = [moduloAB(t, -PI, PI) for t in phi]
        # nr = 100
        theta = np.linspace(phi[0], phi[1], nr)
    return R * np.exp(1j * theta)


def map_data(data_matrix, row_value, ideogram_length):
    n = data_matrix.shape[0]  # square, so same as 1
    mapped = np.zeros([n, n])
    for j in range(n):
        mapped[:, j] = ideogram_length * data_matrix[:, j] / row_value
    return mapped


def make_ribbon_ends(mapped_data, ideo_ends, idx_sort):
    n = mapped_data.shape[0]
    ribbon_boundary = np.zeros((n, n + 1))
    for k in range(n):
        start = ideo_ends[k][0]
        ribbon_boundary[k][0] = start
        for j in range(1, n + 1):
            J = idx_sort[k][j - 1]
            ribbon_boundary[k][j] = start + mapped_data[k][J]
            start = ribbon_boundary[k][j]
    return [
        [(ribbon_boundary[k][j], ribbon_boundary[k][j + 1]) for j in range(n)]
        for k in range(n)
    ]


def control_pts(angle, radius):
    if len(angle) != 3:
        raise ValueError("Angle must have len = 3")
    b_cplx = np.array([np.exp(1j * angle[k]) for k in range(3)])
    b_cplx[1] = radius * b_cplx[1]
    return list(zip(b_cplx.real, b_cplx.imag))


def ctrl_rib_chords(l, r, radius):
    if len(l) != 2 or len(r) != 2:
        raise ValueError("The arc ends must be elements in a list of len 2")
    return [control_pts([l[j], (l[j] + r[j]) / 2, r[j]], radius) for j in range(2)]


def make_q_bezier(b):
    if len(b) != 3:
        raise ValueError("Contaol polygon must have 3 points")
    A, B, C = b
    return (
        "M "
        + str(A[0])
        + ","
        + str(A[1])
        + " "
        + "Q "
        + str(B[0])
        + ", "
        + str(B[1])
        + " "
        + str(C[0])
        + ", "
        + str(C[1])
    )


def make_ribbon_arc(theta0, theta1):
    if test_2PI(theta0) and test_2PI(theta1):
        if theta0 < theta1:
            theta0 = moduloAB(theta0, -PI, PI)
            theta1 = moduloAB(theta1, -PI, PI)
            if theta0 * theta1 > 0:
                raise ValueError("Incorrect angle coordinates for ribbon")
        nr = int(40 * (theta0 - theta1) / PI)
        if nr <= 2:
            nr = 3
        theta = np.linspace(theta0, theta1, nr)
        pts = np.exp(1j * theta)
        string_arc = ""
        for k in range(len(theta)):
            string_arc += "L " + str(pts.real[k]) + ", " + str(pts.imag[k]) + " "
        return string_arc
    else:
        raise ValueError("The angle coords for arc ribbon must be [0, 2*PI]")


def make_layout(title):
    xaxis = dict(
        showline=False, zeroline=False, showgrid=False, showticklabels=False, title="Seatlle Bike Flow"
    )
    yaxis = {**xaxis, "scaleanchor": "x"}
    return dict(
        title=title,
        xaxis=xaxis,
        yaxis=yaxis,
        showlegend=False,
        margin=dict(t=25, b=25, l=25, r=25),
        hovermode="closest",
        shapes=[],
    )


def make_ideo_shape(path, line_color, fill_color):
    return dict(
        line=go.layout.shape.Line(color=line_color, width=0.45),
        path=path,
        type="path",
        fillcolor=fill_color,
        layer="below",
    )


def make_ribbon(l, r, line_color, fill_color, radius=0.2):
    poligon = ctrl_rib_chords(l, r, radius)
    b, c = poligon
    return dict(
        line=go.layout.shape.Line(color=line_color, width=0.5),
        path=make_q_bezier(b)
        + make_ribbon_arc(r[0], r[1])
        + make_q_bezier(c[::-1])
        + make_ribbon_arc(l[1], l[0]),
        type="path",
        fillcolor=fill_color,
        layer="below",
    )


def make_self_rel(l, line_color, fill_color, radius):
    b = control_pts([l[0], (l[0] + l[1]) / 2, l[1]], radius)
    return dict(
        line=dict(color=line_color, width=0.5),
        path=make_q_bezier(b) + make_ribbon_arc(l[1], l[0]),
        type="path",
        fillcolor=fill_color,
        layer="below",
    )


def invPerm(perm):
    inv = [0] * len(perm)
    for i, s in enumerate(perm):
        inv[s] = i
    return inv


def make_filled_chord(M):  # ,labels):

    n = M.shape[0]
    labels = list(M.columns)
    M = M.T
    matrix = M.to_numpy()
    n = M.shape[0]
    row_sum = [np.sum(matrix[k, :]) for k in range(n)]
    gap = 2 * PI * 10e-8

    ideogram_length = 2 * PI * np.asarray(row_sum) / sum(row_sum) - gap * np.ones(n)
    ideo_colors=['rgba(213,62,79,0.75)',
             'rgba(244,109,67,0.75)',
             'rgba(253,174,97,0.75)',
             'rgba(254,224,139,0.75)',
             'rgba(255,255,191,0.75)',
             'rgba(230,245,152,0.75)',
             'rgba(171,221,164,0.75)',
             'rgba(102,194,165,0.75)',
             'rgba(50,136,189,0.75)']
    #ideo_colors = [
    #    x[:3] + "a" + x[3:-1] + ",.75" + x[-1] for x in get_spaced_colors(len(labels))
    #]
    mapped_data = map_data(matrix, row_sum, ideogram_length)
    idx_sort = np.argsort(mapped_data, axis=1)
    ideo_ends = get_ideogram_ends(ideogram_length, gap)
    ribbon_ends = make_ribbon_ends(mapped_data, ideo_ends, idx_sort)
    ribbon_color = [n * [ideo_colors[k]] for k in range(n)]
    layout = make_layout(" ")
    ribbon_info = []
    radii_sribb = [0.2] * n
    for k in range(n):
        sigma = idx_sort[k]
        sigma_inv = invPerm(sigma)
        for j in range(k, n):
            if M.iloc[k, j] == 0 and M.iloc[j, k] == 0:
                continue
            eta = idx_sort[j]
            eta_inv = invPerm(eta)
            l = ribbon_ends[k][sigma_inv[j]]
            if j == k:
                layout["shapes"].append(
                    make_self_rel(
                        l, "rgb(175,175,175)", ideo_colors[k], radius=radii_sribb[k]
                    )
                )
                z = 0.9 * np.exp(1j * (l[0] + l[1]) / 2)
                text = (
                    labels[k]
                    + " \u2b8c "
                    #+ labels[k]
                    + ": {0} viagens ".format(M.iloc[k, k])
                )
                ribbon_info.append(
                    go.Scatter(
                        x=[z.real],
                        y=[z.imag],
                        mode="markers",
                        text=text,
                        hoverinfo="text",
                        marker=dict(size=0.5, color=ideo_colors[k]),
                    )
                )
            else:
                r = ribbon_ends[j][eta_inv[k]]
                zi = 0.9 * np.exp(1j * (l[0] + l[1]) / 2)
                zf = 0.9 * np.exp(1j * (r[0] + r[1]) / 2)

                texti = (
                    labels[k]
                    + " \u2794 "
                    + labels[j]
                    + ": {0} viagens ".format(matrix[k][j])
                )
                textf = (
                    labels[j]
                    + " \u2794 "
                    + labels[k]
                    +": {0} viagens ".format(matrix[j][k])
                )

                ribbon_info.append(
                    go.Scatter(
                        x=[zi.real],
                        y=[zi.imag],
                        mode="markers",
                        text=texti,
                        hoverinfo="text",
                        marker=dict(size=0.5, color=ribbon_color[k][j]),
                    )
                )
                ribbon_info.append(
                    go.Scatter(
                        x=[zf.real],
                        y=[zf.imag],
                        mode="markers",
                        text=textf,
                        hoverinfo="text",
                        marker=dict(size=0.5, color=ribbon_color[j][k]),
                    )
                )
                r = (r[1], r[0])
                if matrix[k][j] > matrix[j][k]:
                    color_of_highest = ribbon_color[k][j]
                else:
                    color_of_highest = ribbon_color[j][k]
                layout["shapes"].append(
                    make_ribbon(l, r, "rgb(175, 175, 175)", color_of_highest)
                )
    ideograms = []
    for k in range(len(ideo_ends)):
        z = make_ideogram_arc(1.1, ideo_ends[k])
        zi = make_ideogram_arc(1.0, ideo_ends[k])
        m = len(z)
        n = len(zi)
        ideograms.append(
            go.Scatter(
                x=z.real,
                y=z.imag,
                mode="lines",
                line=dict(color=ideo_colors[k], shape="spline", width=0.25),
                text=labels[k] + "<br>" + "{0} partidas".format(row_sum[k]),
                hoverinfo="text",
            )
        )
        path = "M "
        for s in range(m):
            path += str(z.real[s]) + ", " + str(z.imag[s]) + " L "
        Zi = np.array(zi.tolist()[::-1])
        for s in range(m):
            path += str(Zi.real[s]) + ", " + str(Zi.imag[s]) + " L "
        path += str(z.real[0]) + " ," + str(z.imag[0])
        layout["shapes"].append(
            make_ideo_shape(path, "rgb(150,150,150)", ideo_colors[k])
        )

    layout["paper_bgcolor"] = "rgba(0,0,0,0)"
    layout["plot_bgcolor"] = "rgba(0,0,0,0)"
    layout["width"] = 625
    layout["height"] = 625
    data = ideograms + ribbon_info
    fig = {"data": data, "layout": layout}
    return fig

In [333]:
chord = make_filled_chord(matrix)

In [334]:
init_notebook_mode(connected=True)
iplot(chord)

Das estações analisadas, a estação Pier 69 é a que tem a maior quantidade de aluguéis de bicicleta (7898). A maior parte das viagens ali iniciadas (53,3%) retorna à própria estação. O segundo maior fluxo observado (2291 viagens) foi entre o Seattle Aquarium e o Pier 69.

Estas duas estações concentram a maior parte das partidas e tem um grande fluxo entre elas.

Pelo gráfico, é possível notar também que grande parte das viagens iniciam e terminam na mesma estação.

In [102]:
df2 = df.loc[df['from_station_name']=='Pier 69 / Alaskan Way & Clay St'].loc[df['to_station_name'].isin(list(union_stations))].reset_index()

In [103]:
df3 = pd.concat([df2,df2['starttime'].str.partition().drop(columns=[1]).rename(columns={0:'date',2:'time'})], axis=1).drop(columns=['index'])

In [104]:
df3

trip_id         starttime          stoptime    bikeid  tripduration  \
0         773  10/13/2014 17:28  10/13/2014 17:52  SEA00445      1438.596   
1         774  10/13/2014 17:28  10/13/2014 17:52  SEA00050      1427.933   
2         777  10/13/2014 17:30  10/13/2014 17:52  SEA00043      1329.031   
3         780  10/13/2014 17:30  10/13/2014 17:52  SEA00263      1290.055   
4         983   10/14/2014 9:12  10/14/2014 13:20  SEA00181     14885.176   
...       ...               ...               ...       ...           ...   
8248   255074   8/31/2016 12:08   8/31/2016 12:24  SEA00266       960.975   
8249   255132   8/31/2016 16:38   8/31/2016 16:51  SEA00261       804.988   
8250   255141   8/31/2016 17:03   8/31/2016 17:12  SEA00449       522.782   
8251   255145   8/31/2016 17:09   8/31/2016 17:20  SEA00185       642.207   
8252   255162   8/31/2016 17:44   8/31/2016 17:49  SEA00455       305.153   

                    from_station_name  \
0     Pier 69 / Alaskan Way & Clay St   
1     Pier 69 / Alaskan Way & Clay St   
2     Pier 69 / Alaskan Way & Clay St   
3     Pier 69 / Alaskan Way & Clay St   
4     Pier 69 / Alaskan Way & Clay St   
...                               ...   
8248  Pier 69 / Alaskan Way & Clay St   
8249  Pier 69 / Alaskan Way & Clay St   
8250  Pier 69 / Alaskan Way & Clay St   
8251  Pier 69 / Alaskan Way & Clay St   
8252  Pier 69 / Alaskan Way & Clay St   

                                        to_station_name from_station_id  \
0                                    3rd Ave & Broad St           WF-01   
1                                    3rd Ave & Broad St           WF-01   
2                                    3rd Ave & Broad St           WF-01   
3                                    3rd Ave & Broad St           WF-01   
4                       Pier 69 / Alaskan Way & Clay St           WF-01   
...                                                 ...             ...   
8248  Seattle Aquarium / Alaskan Way S & Elliott Bay...           WF-01   
8249  King Street Station Plaza / 2nd Ave Extension ...           WF-01   
8250  King Street Station Plaza / 2nd Ave Extension ...           WF-01   
8251  King Street Station Plaza / 2nd Ave Extension ...           WF-01   
8252  Seattle Aquarium / Alaskan Way S & Elliott Bay...           WF-01   

     to_station_id                usertype gender  birthyear        date  \
0            BT-01  Short-Term Pass Holder    NaN        NaN  10/13/2014   
1            BT-01  Short-Term Pass Holder    NaN        NaN  10/13/2014   
2            BT-01  Short-Term Pass Holder    NaN        NaN  10/13/2014   
3            BT-01  Short-Term Pass Holder    NaN        NaN  10/13/2014   
4            WF-01  Short-Term Pass Holder    NaN        NaN  10/14/2014   
...            ...                     ...    ...        ...         ...   
8248         WF-04  Short-Term Pass Holder    NaN        NaN   8/31/2016   
8249         PS-05                  Member   Male     1963.0   8/31/2016   
8250         PS-05                  Member   Male     1980.0   8/31/2016   
8251         PS-05                  Member   Male     1964.0   8/31/2016   
8252         WF-04                  Member   Male     1982.0   8/31/2016   

       time  
0     17:28  
1     17:28  
2     17:30  
3     17:30  
4      9:12  
...     ...  
8248  12:08  
8249  16:38  
8250  17:03  
8251  17:09  
8252  17:44  

[8253 rows x 14 columns]

In [105]:
df3['date'] = pd.to_datetime(df3['date'])

In [106]:
df3['hour'] = pd.to_datetime(df3['time']).dt.hour

In [107]:
df3['day'] = np.where((df3['date']).dt.dayofweek < 5, 'weekday','weekend')

In [108]:
bins = [4, 11, 17, 21]
labels = ['morning','afternoon','night']
df3['time_of_day'] = pd.cut(df3['hour'], bins=bins, labels=labels)

In [109]:
df3['time_of_day'].unique()

['afternoon', 'morning', 'night', NaN]
Categories (3, object): ['morning' < 'afternoon' < 'night']

In [110]:
df3['time_of_day'] = df3['time_of_day'].cat.add_categories('noon').fillna('noon')

In [111]:
df3.isna().sum()

trip_id                 0
starttime               0
stoptime                0
bikeid                  0
tripduration            0
from_station_name       0
to_station_name         0
from_station_id         0
to_station_id           0
usertype                0
gender               6080
birthyear            6080
date                    0
time                    0
hour                    0
day                     0
time_of_day             0
dtype: int64

In [112]:
df3.columns

Index(['trip_id', 'starttime', 'stoptime', 'bikeid', 'tripduration',
       'from_station_name', 'to_station_name', 'from_station_id',
       'to_station_id', 'usertype', 'gender', 'birthyear', 'date', 'time',
       'hour', 'day', 'time_of_day'],
      dtype='object')

In [208]:
df4 = pd.concat([df3['from_station_name'],df3['day'],df3['time_of_day'],df3['to_station_name']], axis=1)

In [209]:
df4.replace({'from_station_name': name_dict, 'to_station_name':name_dict}, inplace=True)

In [212]:
df4

from_station_name      day time_of_day            to_station_name
0              Pier 69  weekday   afternoon         3rd Ave & Broad St
1              Pier 69  weekday   afternoon         3rd Ave & Broad St
2              Pier 69  weekday   afternoon         3rd Ave & Broad St
3              Pier 69  weekday   afternoon         3rd Ave & Broad St
4              Pier 69  weekday     morning                    Pier 69
...                ...      ...         ...                        ...
8248           Pier 69  weekday   afternoon           Seattle Aquarium
8249           Pier 69  weekday   afternoon  King Street Station Plaza
8250           Pier 69  weekday   afternoon  King Street Station Plaza
8251           Pier 69  weekday   afternoon  King Street Station Plaza
8252           Pier 69  weekday   afternoon           Seattle Aquarium

[8253 rows x 4 columns]

In [116]:
df5 = df4.groupby(['from_station_name', 'day','time_of_day','to_station_name']).size().reset_index(name='count')

In [117]:
df5

from_station_name      day time_of_day            to_station_name  count
0            Pier 69  weekday     morning          2nd Ave & Pine St     27
1            Pier 69  weekday     morning         3rd Ave & Broad St     45
2            Pier 69  weekday     morning         Burke-Gilman Trail      0
3            Pier 69  weekday     morning       E Pine St & 16th Ave      0
4            Pier 69  weekday     morning  King Street Station Plaza     19
..               ...      ...         ...                        ...    ...
67           Pier 69  weekend        noon  King Street Station Plaza      0
68           Pier 69  weekend        noon            Lake Union Park      0
69           Pier 69  weekend        noon                    Pier 69     37
70           Pier 69  weekend        noon          Pine St & 9th Ave      1
71           Pier 69  weekend        noon           Seattle Aquarium     16

[72 rows x 5 columns]

In [156]:
teste= df2.groupby(['from_station_name','to_station_name']).size().reset_index(name='count')

In [157]:
teste.replace({'from_station_name': name_dict, 'to_station_name':name_dict}, inplace=True)

In [167]:
teste['to_station_name'].unique()

array(['2nd Ave & Pine St', '3rd Ave & Broad St', 'Burke-Gilman Trail',
       'E Pine St & 16th Ave', 'King Street Station Plaza',
       'Lake Union Park', 'Pier 69', 'Pine St & 9th Ave',
       'Seattle Aquarium'], dtype=object)

In [168]:
teste['to_station_name'] = teste['to_station_name'].astype('category')
teste['to_station_cat'] = teste['to_station_name'].cat.codes

In [170]:
teste['from_station_cat'] = 6

In [171]:
teste

from_station_name            to_station_name  count  to_station_cat  \
0           Pier 69          2nd Ave & Pine St    310               0   
1           Pier 69         3rd Ave & Broad St    347               1   
2           Pier 69         Burke-Gilman Trail      4               2   
3           Pier 69       E Pine St & 16th Ave      3               3   
4           Pier 69  King Street Station Plaza    962               4   
5           Pier 69            Lake Union Park     75               5   
6           Pier 69                    Pier 69   4209               6   
7           Pier 69          Pine St & 9th Ave     52               7   
8           Pier 69           Seattle Aquarium   2291               8   

   from_station_cat  
0                 6  
1                 6  
2                 6  
3                 6  
4                 6  
5                 6  
6                 6  
7                 6  
8                 6

In [174]:
source = teste['from_station_cat']
target = teste['to_station_cat']
value = teste['count']
label = teste['to_station_name']
link = dict(source = source, target = target, value = value)
node = dict(label = label)
data = go.Sankey(link = link, node=node)

fig_teste = go.Figure(data)
init_notebook_mode(connected=True)
iplot(fig_teste)

In [179]:
station_time = df4.groupby(['from_station_name', 'time_of_day']).size().reset_index(name='count_1')

In [180]:
station_time['from_station_cat'] = 0
station_time['time_of_day_cat'] = pd.factorize(station_time['time_of_day'].sort)[0] + 1

In [213]:
station_time

from_station_name time_of_day  count_1  from_station_cat  time_of_day_cat
0           Pier 69     morning     1636                 0                1
1           Pier 69   afternoon     5066                 0                2
2           Pier 69       night     1379                 0                3
3           Pier 69        noon      172                 0                4

In [199]:
label = list(np.unique(station_time[['from_station_name','time_of_day']].values))

array(['Pier 69', 'afternoon', 'morning', 'night', 'noon'], dtype=object)

In [214]:
station_time

from_station_name time_of_day  count_1  from_station_cat  time_of_day_cat
0           Pier 69     morning     1636                 0                1
1           Pier 69   afternoon     5066                 0                2
2           Pier 69       night     1379                 0                3
3           Pier 69        noon      172                 0                4

In [215]:
source = station_time['from_station_cat']
target = station_time['time_of_day_cat']
value = station_time['count_1']
label = ['Pier 69','morning','afternoon','night','noon']
#list(np.unique(station_time[['from_station_name','time_of_day']].values))
link = dict(source = source, target = target, value = value)
node = dict(label = label)
data = go.Sankey(link = link, node=node)

fig_teste = go.Figure(data)
init_notebook_mode(connected=True)
iplot(fig_teste)

In [247]:
time_weekday = df4.groupby(['time_of_day','day']).size().reset_index(name='count_2')

In [248]:
time_weekday

time_of_day      day  count_2
0     morning  weekday     1034
1     morning  weekend      602
2   afternoon  weekday     3185
3   afternoon  weekend     1881
4       night  weekday      924
5       night  weekend      455
6        noon  weekday      109
7        noon  weekend       63

In [249]:
df=station_time[['time_of_day','time_of_day_cat']].T
df.columns = df.iloc[0]
df = df[1:]
dict_time = df.to_dict('records')

In [250]:
dict_time[0]

{'morning': 1, 'afternoon': 2, 'night': 3, 'noon': 4}

In [251]:
time_weekday['day_cat'] = time_weekday['day'].map({'weekday': 5, 'weekend': 6})
time_weekday['time_of_day_cat'] = time_weekday['time_of_day'].map(dict_time[0])
time_weekday

time_of_day      day  count_2  day_cat time_of_day_cat
0     morning  weekday     1034        5               1
1     morning  weekend      602        6               1
2   afternoon  weekday     3185        5               2
3   afternoon  weekend     1881        6               2
4       night  weekday      924        5               3
5       night  weekend      455        6               3
6        noon  weekday      109        5               4
7        noon  weekend       63        6               4

In [270]:
source = station_time['from_station_cat'].append(time_weekday['time_of_day_cat'])
target = station_time['time_of_day_cat'].append(time_weekday['day_cat'])
value = station_time['count_1'].append(time_weekday['count_2'])
label = ['Pier 69','morning','afternoon','night','noon','weekday','weekend']

In [271]:
link = dict(source = source, target = target, value = value)
node = dict(label = label)
data = go.Sankey(link = link, node=node)

fig_teste2 = go.Figure(data)
init_notebook_mode(connected=True)
iplot(fig_teste2)

In [185]:
weekday_station = df4.groupby(['day','to_station_name']).size().reset_index(name='count_3')

In [186]:
weekday_station

day            to_station_name  count_3
0   weekday          2nd Ave & Pine St      223
1   weekday         3rd Ave & Broad St      183
2   weekday         Burke-Gilman Trail        2
3   weekday       E Pine St & 16th Ave        2
4   weekday  King Street Station Plaza      927
5   weekday            Lake Union Park       34
6   weekday                    Pier 69     2318
7   weekday          Pine St & 9th Ave       33
8   weekday           Seattle Aquarium     1530
9   weekend          2nd Ave & Pine St       87
10  weekend         3rd Ave & Broad St      164
11  weekend         Burke-Gilman Trail        2
12  weekend       E Pine St & 16th Ave        1
13  weekend  King Street Station Plaza       35
14  weekend            Lake Union Park       41
15  weekend                    Pier 69     1891
16  weekend          Pine St & 9th Ave       19
17  weekend           Seattle Aquarium      761

In [272]:
weekday_station['day_cat'] = weekday_station['day'].map({'weekday': 5, 'weekend': 6})

In [274]:
dict_station = {'Pier 69':0,'2nd Ave & Pine St':7, '3rd Ave & Broad St':8, 'Burke-Gilman Trail':9,
                'E Pine St & 16th Ave':10, 'King Street Station Plaza':11,'Lake Union Park':12,
                'Pine St & 9th Ave':13,'Seattle Aquarium':14}

In [276]:
weekday_station['to_station_cat'] = weekday_station['to_station_name'].map(dict_station)

In [277]:
weekday_station

day            to_station_name  count_3  day_cat  to_station_cat
0   weekday          2nd Ave & Pine St      223        5               7
1   weekday         3rd Ave & Broad St      183        5               8
2   weekday         Burke-Gilman Trail        2        5               9
3   weekday       E Pine St & 16th Ave        2        5              10
4   weekday  King Street Station Plaza      927        5              11
5   weekday            Lake Union Park       34        5              12
6   weekday                    Pier 69     2318        5               0
7   weekday          Pine St & 9th Ave       33        5              13
8   weekday           Seattle Aquarium     1530        5              14
9   weekend          2nd Ave & Pine St       87        6               7
10  weekend         3rd Ave & Broad St      164        6               8
11  weekend         Burke-Gilman Trail        2        6               9
12  weekend       E Pine St & 16th Ave        1        6              10
13  weekend  King Street Station Plaza       35        6              11
14  weekend            Lake Union Park       41        6              12
15  weekend                    Pier 69     1891        6               0
16  weekend          Pine St & 9th Ave       19        6              13
17  weekend           Seattle Aquarium      761        6              14

In [278]:
source = station_time['from_station_cat'].append(time_weekday['time_of_day_cat']).append(weekday_station['day_cat'])
target = station_time['time_of_day_cat'].append(time_weekday['day_cat']).append(weekday_station['to_station_cat'])
value = station_time['count_1'].append(time_weekday['count_2']).append(weekday_station['count_3'])
label = ['Pier 69','morning','afternoon','night','noon','weekday','weekend','2nd Ave & Pine St', '3rd Ave & Broad St',
         'Burke-Gilman Trail','E Pine St & 16th Ave', 'King Street Station Plaza','Lake Union Park','Pine St & 9th Ave','Seattle Aquarium']

In [314]:
color_node = ['#756bb1',
              '#e6550d', '#fee391','#3182bd','#000000',
              '#de77ae', '#238b45', 
              '#1b9e77', '#d95f02','#7570b3', '#e7298a', '#66a61e', '#e6ab02', '#756bb1','#a6761d', '#666666']

color_link = ['#bcbddc', '#bcbddc', '#bcbddc', '#bcbddc',
              '#fdae6b', '#fdae6b','#fff7bc', '#fff7bc','#9ecae1', '#9ecae1', '#525252', '#525252',
              '#fa9fb5', '#fa9fb5','#fa9fb5', '#fa9fb5', '#fa9fb5', '#fa9fb5', '#fa9fb5','#fa9fb5', '#fa9fb5',
              '#a1d99b', '#a1d99b','#a1d99b', '#a1d99b', '#a1d99b', '#a1d99b', '#a1d99b','#a1d99b', '#a1d99b']

In [322]:
link = dict(source = source, target = target, value = value, color=color_link)
node = dict(label = label,color=color_node)
data = go.Sankey(link = link, node=node)

sankey = go.Figure(data)
sankey.update_layout(title_text='Seattle Bicycle Trips Flow from Pier 69 Station',font_size=14)
init_notebook_mode(connected=True)
iplot(sankey)

A análise do diagrama Sankey do fluxo de viagens de bicicleta a partir da estação Pier 69 revela que a maior parte das viagens se inicia entre 12h e 18h, durante a semana. Há uma pequena redução da quantidade de viagens que se iniciam e terminam na estação analisada durante os finais de semana, porém este movimento segue sendo extremamente relevante, o que não se observa no fluxo entre esta estação e as demais.